pyrealm/demography/canopy.py

```diff
@@ -0,0 +1,149 @@
+"""Functionality for canopy modelling."""
+
+import numpy as np
+from numpy.typing import NDArray
+from scipy.optimize import root_scalar  # type: ignore [import-untyped]
+
+from pyrealm.demography.canopy_functions import (
+    calculate_relative_canopy_radius_at_z,
+    calculate_stem_projected_crown_area_at_z,
+    calculate_stem_projected_leaf_area_at_z,
+    solve_community_projected_canopy_area,
+)
+from pyrealm.demography.community import Community
+
+
+class Canopy:
+    """Model of the canopy for a plant community.
+
+    This class generates a canopy structure for a community of trees using the
+    perfect-plasticity approximation model :cite:`purves:2008a`. In this approach, each
+    individual is assumed to arrange its canopy crown area plastically to take up space
+    in canopy layers and that new layers form below the canopy top as the available
+    space is occupied.
+
+    Real canopies contain canopy gaps, through process such as crown shyness. This
+    is included in the model through the canopy gap fraction, which sets the proportion
+    of the available space that will remain unfilled by any canopy.
+
+    Args:
+        community: A Community object that will be used to generate the canopy model.
+        canopy_gap_fraction: The proportion of the available space unfilled by canopy
+            (default: 0.05).
+        layer_tolerance: The minimum precision used by the solver to find canopy layer
+            closure heights (default: 0.001 metres)
+    """
+
+    def __init__(
+        self,
+        community: Community,
+        canopy_gap_fraction: float = 0.05,
+        layer_tolerance: float = 0.001,
+    ) -> None:
+        # Calculate community wide properties: total crown area and maximum height
```

Comment:
This code is very strongly tied to the `__init__` method. It _would_ look cleaner to declare the attributes in the `__init__` and then have a separate method for the maths, but that seems almost entirely cosmetic. And I don't mean to knock cosmetic at all - clarity matters. I just can't see how this leads to meaningful separation of concerns. I guess I'm mostly thinking testing - the `__init__` and `do_maths` would be largely inseparable for testing purposes?

Reply:
Done - I think it is definitely cleaner but still curious if it offers anything else in this case.

Reply:
It's definitely cleaner. I think ideally, I'd break out the root finding loop into a function to be called from `canopy_functions.py` but it's not a deal breaker at all. It would be more testable, and if you're thinking of replacing it with another root finding algorithm, more modular.

This class to me is essentially a dataclass, and where we've implemented data classes previously e.g. `calendar` we've used post-init methods to process the inputs and provide values for attributes. I like this style and it keeps things tidy.

This is all ultimately aesthetic preference. There's not really a computational implication.

Reply:
There can be some computational advantages (may not be applicable here) to what James is suggesting. If you can separate out some of the `__init__` calculations into separate functions and wrap them in `@property` decorator then you can benefit from lazy evaluation.

Reply:
Yeah - this could very well be a dataclass. I don't think it's an issue to leave as is for the moment but I see your point about the structure. I'll leave the root finding loop as is for now. 

@omarjamil That's a good point about lazy evaluation - I don't think it applies here because we need all of the attributes calculated to move on to the next step but that it is something I'll keep in mind in future.